core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -715,12 +715,8 @@ private[spark] object Utils extends Logging {
 
   /** Get the Yarn approved local directories. */
   private def getYarnLocalDirs(conf: SparkConf): String = {
-    // Hadoop 0.23 and 2.x have different Environment variable names for the
-    // local dirs, so lets check both. We assume one of the 2 is set.
-    // LOCAL_DIRS => 2.X, YARN_LOCAL_DIRS => 0.23.X
-    val localDirs = Option(conf.getenv("YARN_LOCAL_DIRS"))
-      .getOrElse(Option(conf.getenv("LOCAL_DIRS"))
-      .getOrElse(""))
+    //YarnLocalDirs must be inside container directory. Since it will be automatically deleted when container shut downs.
+    val localDirs = Option(System.getProperty("user.dir")).getOrElse(""))
```

Comment:
So e.g. in spark run via yarn-client
i can see directory structures like 
{yarn.nodemanager.local-dirs}/nm-local-dir/usercache/admin/appcache/application_1424859293845_0003/container_1424859293845_0003_01_000001/ -- this is current working directory since executor was launched from this directory.
and spark is using {yarn.nodemanager.local-dirs}/nm-local-dir/usercache/admin/appcache/application_1424859293845_0003/ -- this directory to write shuffle files which will gets deleted when application shuts down.
And also regarding  #4759 (comment) will not work if executor gets killed without letting shutdown hook to trigger i.e.

-pankaj